parsers/verilog.c

```diff
@@ -1171,27 +1172,66 @@ static int skipParameterAssignment (int c)
 static int processFunction (tokenInfo *const token, int c)
 {
 	verilogKind kind = token->kind;	// K_FUNCTION or K_TASK
+#define TOKEN_NAME_SIZE 16
+	vString *tokenNameSaved[TOKEN_NAME_SIZE];
+	int doubleColonPos[TOKEN_NAME_SIZE];
+	int functionNameStart = 0;
+	int i, j;
+
+	for (i = 0; i < TOKEN_NAME_SIZE; i++)
+		tokenNameSaved[i] = vStringNew ();
 
 	/* Search for function name
 	 * Last identifier found before a '(' or a ';' is the function name */
+	i = 0;
+	j = 0;
 	while (c != '(' && c != ';' && c != EOF)
 	{
 		if (isWordToken (c))
+		{
 			c = readWordToken (token, c);
+			if (i < TOKEN_NAME_SIZE)
+				vStringCopy (tokenNameSaved[i++], token->name);
+		}
 		else
 			c = skipWhite (vGetc ());
 		/* skip parameter assignment of a class type
 		 *	ex. function uvm_port_base #(IF) get_if(int index=0); */
 		c = skipParameterAssignment (c);
 
-		/* Identify class type prefixes and create respective context*/
-		if (isInputLanguage (Lang_systemverilog) && isDoubleColon(c))
+		/* Record the position where the class type prefix appears */
+		if (isInputLanguage (Lang_systemverilog) && isDoubleColon(c) && j < TOKEN_NAME_SIZE)
+			doubleColonPos[j++] = i;
+	}
+
+	/* Search for class_scope of function name and create respective context */
+	if (j == 1 && i > 1 && (i - doubleColonPos[0]) == 1)
+		functionNameStart = i - 1;
+	else if (j > 1)
+	{
+		for (int k = j - 1; k >= 1; k--)
+			if (doubleColonPos[k] - doubleColonPos[k-1] > 1)
+			{
+				functionNameStart = doubleColonPos[k];
+				break;
+			}
+		if (functionNameStart == 0 && (i - doubleColonPos[j-1]) == 1)
+			functionNameStart = doubleColonPos[0];
+	}
```

Comment:
These line for deciding functionNameStart.
What do you think about converting these lines to a function that returns funtionNameStart like:
```
functionNameStart = calcFunctionNameStart(tokenNameSaved, doubleColonPos);
```

Reply:
Okay, but according to the Verilog language LRM, this function is unlikely to be used in other situations. I can do this and add some comments to this.